Add option to split Linear gates for Quantizable LSTM into separate ops

WongJohnson opened the pull request:
Summary:
For LSTM, the input and hidden state are projected with Linear layers to construct the 4 gates. This is typically performed together as a single Linear (for each state) with output channel count `4 * hidden_dim` for efficiency. 
https://www.internalfb.com/code/fbsource/[ebef7c4238aa55948b2b444044f2c8ed2040de55]/fbcode/caffe2/torch/ao/nn/quantizable/modules/rnn.py?lines=52-58
The output is then ultimately split into 4:
https://www.internalfb.com/code/fbsource/[ebef7c4238aa55948b2b444044f2c8ed2040de55]/fbcode/caffe2/torch/ao/nn/quantizable/modules/rnn.py?lines=83-87

For on-device latency (and possibly memory) considerations, we want to avoid constructing the intermediate `gates` tensor (which can be relatively large), by splitting `igates` and `hgates` first (as 4x `Linear(hidden_dim, hidden_dim)` each), applying add separately, then proceeding as usual.

This functionality can be enabled by specifying `split_gates=True` (default False is original behavior) at any entry point (directly with `torch.ao.nn.quantizable.LSTM`  or via `_get_lstm_with_individually_observed_parts`).

Test Plan:
piggy back on existing test to check for correct swap handling, numerics, and jit.script during prepare/convert
```
buck2 test 'fbcode//mode/opt' fbcode//caffe2/test/quantization:test_quantization -- --exact 'caffe2/test/quantization:test_quantization - test_custom_module_lstm (caffe2.test.quantization.core.test_quantized_op.TestQuantizedOps)'
```
https://www.internalfb.com/intern/testinfra/testrun/11540474102848372

This test is quite long running now (more than double original).

Reviewed By: Ninja91

Differential Revision: D65283170




cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv

jerryzh168 commented on test/quantization/core/test_quantized_op.py:
```diff
@@ -2917,6 +2917,11 @@ def test_advanced_indexing(self):
 
     @override_qengines
     def test_custom_module_lstm(self):
+        class QuantizedLSTMSplitGates(torch.ao.nn.quantizable.LSTM):
```
for UX, will it be easier to reuse the default mapping: https://github.com/pytorch/pytorch/blob/e39955e82f494dcf6f22c0d81556e7cc26a281f7/torch/ao/quantization/quantize.py#L54 and just have user set some attributes on `nn.LSTM`?

jerryzh168 replied:
also this class should be called `QuantizableLSTMSplitGates` btw if you want to use this

WongJohnson replied:
> just have user set some attributes on nn.LSTM

I thought about that but since we are not changing any behavior in `nn.LSTM` I was hesitant to change any code there. Definitely open to trying a cleaner way to toggle this if something comes up!

WongJohnson committed:
Add option to split Linear gates for Quantizable LSTM into separate ops (#140868)

Summary:

For LSTM, the input and hidden state are projected with Linear layers to construct the 4 gates. This is typically performed together as a single Linear (for each state) with output channel count `4 * hidden_dim` for efficiency. 
https://www.internalfb.com/code/fbsource/[ebef7c4238aa55948b2b444044f2c8ed2040de55]/fbcode/caffe2/torch/ao/nn/quantizable/modules/rnn.py?lines=52-58
The output is then ultimately split into 4:
https://www.internalfb.com/code/fbsource/[ebef7c4238aa55948b2b444044f2c8ed2040de55]/fbcode/caffe2/torch/ao/nn/quantizable/modules/rnn.py?lines=83-87

For on-device latency (and possibly memory) considerations, we want to avoid constructing the intermediate `gates` tensor (which can be relatively large), by splitting `igates` and `hgates` first (as 4x `Linear(hidden_dim, hidden_dim)` each), applying add separately, then proceeding as usual.

This functionality can be enabled by specifying `split_gates=True` (default False is original behavior) at any entry point (directly with `torch.ao.nn.quantizable.LSTM`  or via `_get_lstm_with_individually_observed_parts`).

Test Plan:
piggy back on existing test to check for correct swap handling, numerics, and jit.script during prepare/convert
```
buck2 test 'fbcode//mode/opt' fbcode//caffe2/test/quantization:test_quantization -- --exact 'caffe2/test/quantization:test_quantization - test_custom_module_lstm (caffe2.test.quantization.core.test_quantized_op.TestQuantizedOps)'
```
https://www.internalfb.com/intern/testinfra/testrun/11540474102848372

This test is quite long running now (more than double original).

Reviewed By: Ninja91

Differential Revision: D65283170

WongJohnson commented:
@pytorchmergebot merge

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change but I think lint is failing on this in trunk' -c weird

[GH job link](https://github.com/pytorch/pytorch/actions/runs/11966726863/job/33362825428) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/3acc6eac492dd93ae4c1dc202bdf98ef85418a1b)

This is weird because the PR looks fine and the lint failure only shows up flakily in trunk.  Please do a rebase and reland.
